### ISSUE TITLE
Stuttering playback on many systems with audio set to 48 kHz

### ISSUE THREAD

**Issue by MarcSabatella** [musescore/MuseScore#18701]
### Issue type

General playback bug

### Bug description

On many systems, playback of MuseScore on any non-trivial score is extremely choppy, and somehow, just going to the device settings and changing the audio sample rate from 48 to 44.1 kHz almost always fixes it.  The improvement is far too dramatic to be explained by 44.1 being ever so slightly less demanding on CPU/RAM than 48 - I think something is going more significantly wrong.

@bkunda I don't know how much exposure this problem has within the core team, but on the forums and in social media, it's an enormous black eye.   It's no exaggeration to say we get several reports per day of people with similar audio issues, and several times a day we tell them to try changing their sample rate to 44.1, and several times a day they report back that it worked.  Of course, we also get many people ranting about how terrible the playback is and just assuming it's that MU4 is inherently incapable of sounding good.

I suspect there is something going on where something is expecting 44.1 and something else is providing 48 (or vice versa), and in the attempt to reconcile the two through resampling or whatever, the sound gets chopped up.

On the off chance that this turns out to be something really simple like a missing parameter in a function call somewhere to honor the device sample rate directly and avoid extra conversions automatically, it would be awesome if someone could look into this!

### Steps to reproduce

1. Try to play any reasonably large score (especially with Muse Sounds but I am not sure that it's exclusive to that)
2. In your OS settings, try setting sample rate to 48 vs 44.1 kHz

Result: on many systems, 44.1 sounds normal, 48 sounds terrible

### Screenshots/Screen recordings

(I'll try to get screen recordings later from a system showing this)

### MuseScore Version

OS: Debian GNU/Linux 11 (bullseye), Arch.: x86_64, MuseScore version (64-bit): 4.1.0-231921359, revision: github-musescore-musescore-2e3a93a

### Regression

Yes, this used to work in MuseScore 3.x and now is broken

### Operating system

Windows, Linux

### Additional context

On the systems where the problem occurs, it occurs in all MU4 versions.  The new reverb in 4.1 seems to help some, but still, 44.1 kHz performs much better than 48 on many systems - too much better to just be a matter of the slight reduction in the sample rate itself.

So far the vast majority of the reports are on Windows, which makes sense since that's where most users are.  But also quite a few on Linux, and relatively few on macOS, which is interesting in itself.  But conversely, reports of extreme lag between note input and sound being heard are more common on macOS, and I now wonder if this is somehow related.

There are some other issues relating to specific problems with sample rate settings (see for instance #18579, #17877, #17849, and especially #17313), and it's possible they will all turn out to have the same root cause, but right now there seems to be no other issue specifically covering a general issue with 48 kHz playback that is solved by changing to 44.1.

**Comment by MarcSabatella** [comment]
Yeah, it definitely doesn't happen on all systems.  But it's common enough - it literally comes up several times a day - that it does seem to indicate a real problem to me.  Might be specific to some particular audio device/driver, might relate to some specific Windows patch or some particular Linux audio subsystem, etc.  I recognize it's next to impossible to investigate without a specific system to reproduce it on, and if this were just a sporadic thing, I wouldn't bother opening an issue.  But it's truly pretty pervasive in the real world and is harming our reputation (the social media rants get pretty out of control), so I wanted to make sure we had a place to add information.  I'll try to collect screen recordings and volunteers experiencing the problem who can run diagnostics etc.

**Comment by bkunda** [comment]
I'd be interested to see if issues concerning playback at 48kHz start to become clearer once we start resolving the problems we know how to reproduce (e.g. https://github.com/musescore/MuseScore/issues/17877, which we've triaged as a P0 to get things moving asap).

**Comment by bluebear94** [comment]
If I add a log message in `AudioModule::setupAudioWorker` printing `activeSpec.sampleRate`, then it prints `44100` for me regardless of the current sample rate. Hard-coding 48000 for the parameter to the `AudioEngine::setSampleRate` call doesn’t fix the stutter when the sample rate is 48 kHz (and doesn’t cause the stutter when it’s 44.1 kHz, though it does stretch playback).

**Comment by rtbo** [comment]
I have the issue on my machine, running Ubuntu and Pipewire.
I reproduce the choppy playback as soon as MS sample rate is different than pipewire sample rate.
I tested 4 configurations:

| Pipewire| MuseScore | Playback |
|---|---|---|
| 44100 | 44100 | OK |
| 44100 | 48000 | choppy |
| 48000 | 44100 | choppy |
| 48000 | 48000 | OK |

**Comment by rtbo** [comment]
I don't know how to detect the "correct" config.
In `LinuxAudioDriver::open`, I added checks with `snd_pcm_hw_params_test_rate`.
It reports that both `44100` and `48000` are supported.
Which should be somehow. ALSA is normally able to resample to the correct rate.

**Comment by MarcSabatella** [comment]
It's definitely still an issue on Linux (4.3.2 and master), *if* audio is set up to only allow 48 kHz (which is the default on my system).  I recently discovered that when using pipewire, all works fine if I set clock.allowed-rates to '[ 44100 48000 ]' - which is to say, if it is set up to allow applications to change the sample rate to 44.1 themselves, which is what I assume MuseScore Studio is trying to do.  So that's my workaround currently - I leave the system rate at 48 but include 44.1 in the "allowed" rates so MuseScore Studio can set it that way (and presumably other programs could set it differently).  Previously I was using clock.force-rate to force the rate to 44.1 system-wide.  I assume other Linux audio systems have similar settings, but I'm not really an expert on how all that works.

Now, in current nightly builds from master, there is a new and significantly worse behavior, but with something of a silver lining.  If I return my system to its default settings - 48 kHz system-wide, no 44.1 in the allowed rates - then not only is MuseScore Studio playback choppy, but also too fast and too high.  4.3.2 doesn't do that - the choppiness is the only symptom, and only in "big" scores, making it possible to believe it's a CPU or RAM or buffer size issue instead of a sample rate issue.  The good ppsrt of this is, with 4.4 iut's more obvious exactly what is going on.

I notice there is a sample rate setting with MuseScore Studio that I can access via the DevTools panel, but the spinbox only accepts values up to 999.  If I edit the ".ini" file directly, though, I can confirm as others have reported that setting 48000 allows it to work well using the default audio settings on my system.

As for Windows, we do still get reports pretty often about stuttering audio, and most of the time when we suggest changing audio sample rate, people report that fixes it - if they are able to figure out how to do it at all.  Some people report that their audio device doesn't support any other rate but 48 kHz, and they often end up just giving up on MU4.

The reports of stuttering audio used to happen several times per day - literally hundreds of such reports over the first few months of MU4, on the forums, social media, etc.  Now it's more like a couple per week.  Hard to say if that slowdown is because 4.3.2 is better than 4.0 was, or just because there are fewer new users of MU4 than there were a year ago.

I don't have any issue with this on my own Windows computer, so I can't say if 4.4 works any better than 4.3.2 on the systems that are affected.  I can try some more testing with a couple of other audio devices - but not sure when I'll be able to do.

**Comment by MarcSabatella** [comment]
I would add that independently of which systems show the problem with which audio systems, we can make all of our lives much simpler by exposing (and fixing the spinbox for) the sample rate setting within MuseScore Studio.  Right now when the problem happens, it's really difficult to help people find the appropriate settings on their systems, and as I said, sometimes they hit a dead end (particularly Bluetooth devices often seen locked to 48 kHz).  If we could just send them to Edit / Preferences we could make a whole lot of people much happier much more easily.

If that also allowed people to set 96 kHz - a very common request as well for people with external audio devices - then even better.  That normally causes MuseScore Studio audio to play double speed an octave too high, as well as being choppy.

**Comment by jeetee** [comment]
Looks like all from 44.1 to 192 is expected to be(come?) supported.
For reference, this is the setting with the .ini key: https://github.com/search?q=repo%3Amusescore%2FMuseScore+AUDIO_SAMPLE_RATE_KEY&type=code

**Comment by rtbo** [comment]
I've tracked down on my system that this issue is caused by the [Ubuntu Studio Audio Configuration tool](https://ubuntustudio.org/audio-configuration/) which sets system-wide the `PIPEWIRE_QUANTUM` environment variable.
e.g. `PIPEWIRE_QUANTUM="1024/48000" will **force** pipewire to convert ALSA output to 1024 samples at 48 kHz.

One having the same problem can delete this file: `/etc/profile.d/ubuntustudio-pwjack.sh`.
Or `unset` this variable from a terminal before starting MuseScore.

I guess unsetting from the application (one of the first thing in `main`) would also solve the issue.

**Comment by bkunda** [comment]
Partially fixed by https://github.com/musescore/MuseScore/pull/23954, and also by #23736. 

For the rest, we'll need to allocate time in another release cycle.

**Comment by jamshark70** [comment]
> e.g. PIPEWIRE_QUANTUM="1024/48000" will force pipewire to convert ALSA output to 1024 samples at 48 kHz.
> 
> One having the same problem can delete this file: /etc/profile.d/ubuntustudio-pwjack.sh.
> Or unset this variable from a terminal before starting MuseScore.

Unsetting this variable when pipewire sample rate = 48kHz did *not* resolve the issue for me.

Using Ubuntu Studio Audio Config tool to set 1024 frames (or even 256 frames), 44100 kHz *did* work.

When using a pure JACK config, no Pipewire, I could not use the built-in audio at 44.1 kHz, so I suppose Pipewire is doing some upsampling in there. So it would be nice if MuseScore "just worked" at the user's choice of SR.

**Comment by DmitryArefiev** [comment]
@MarcSabatella @jamshark70 Can you try the build from #23939 on you Linux system please?

**Comment by MarcSabatella** [comment]
Nice, seems to play smoothly at 44.1 or 48, even with a 1024 buffer size.  Whereas the previous nightly build still stutters at 44.1.

FWIW, I am not sure what my audio configuration is.  It's whatever the default is for the Debian container set up within ChromeOS.  I know /usr/bin/pipewire is running and I am able to use some Pipewire tools, but I think the system is not really running Pipewire fully.  I am pretty sure ALSA is involved as well, but again, maybe not fully.  Whatever it is, it's probably actually just a front end for the ChromeOS audio system, which interfaces with Linux via a subsystem called "cras" (Chromium audio server).  Or at least that is how it used to work.

**Comment by jamshark70** [comment]
@DmitryArefiev 
> Can you try the build from https://github.com/musescore/MuseScore/pull/23939 on you Linux system please?

Apologies, I overlooked this... was on holiday at that time, and then broke my leg 😮 literally just 2 days after you asked me to check this and I never got back to it.

Found the build and I'll run it for a few days, see how it goes. The issue happens quite frequently in the production build so I think it shouldn't take long to check it.

**Comment by jamshark70** [comment]
@DmitryArefiev AFAICS the new pipewire driver is working cleanly -- several hours' usage with no failures. Thanks!

**Comment by jamshark70** [comment]
However -- is there any chance that anything in #23939 breaks MIDI file export?

I just tried to export a part from a score to MIDI -- using the pipewire test build, no notes were exported. Using the production 4.5.2 build, MIDI export was fine.

Test build, could be anything -- just reporting this odd behavior with it.

**Comment by cbjeukendrup** [comment]
There was a time where midi export was indeed broken in development builds. That PR is probably from that time, and has not been rebased since then, so the build doesn't contain the fix for that problem yet. But I don't think the PR itself is the cause.

**Comment by sebicodes** [comment]
It is a fluctuating error, it has obviously something to do with the 48 / 44.1 kHz issue and resampling. On my system with ubuntu noble numbat I am not even able to playback a single note without stuttering.

As I am able to reconstruct: 
 MuseScore-Studio-4.6.0.252301442-x86_64.AppImage Not working (as far as I remember)
 MuseScore-Studio-4.6.0.252730944-x86_64.AppImage Working (tested today)
 MuseScore-Studio-4.6.3.252940956-x86_64.AppImage Working (tested today) 
 MuseScore-Studio-4.6.5.253511702-x86_64.AppImage Not working (tested today)

As a workaround I change the pipewire settings manually from the command line:
These are the default settings in my system:
$ pw-metadata -n settings
Found "settings" metadata 31
update: id:0 key:'log.level' value:'2' type:''
update: id:0 key:'clock.rate' value:'48000' type:''
update: id:0 key:'clock.allowed-rates' value:'[ 48000 ]' type:''
update: id:0 key:'clock.quantum' value:'1024' type:''
update: id:0 key:'clock.min-quantum' value:'32' type:''
update: id:0 key:'clock.max-quantum' value:'2048' type:''
update: id:0 key:'clock.force-quantum' value:'1024' type:''
update: id:0 key:'clock.force-rate' value:'48000' type:''

It is sufficient to change the key clock.force-rate to 44100, then I get rid of the stuttering.
$ pw-metadata -n settings 0 clock.force-rate 44100
$ pw-metadata -n settings
Found "settings" metadata 31
update: id:0 key:'log.level' value:'2' type:''
update: id:0 key:'clock.rate' value:'48000' type:''
update: id:0 key:'clock.allowed-rates' value:'[ 48000 ]' type:''
update: id:0 key:'clock.quantum' value:'1024' type:''
update: id:0 key:'clock.min-quantum' value:'32' type:''
update: id:0 key:'clock.max-quantum' value:'2048' type:''
update: id:0 key:'clock.force-quantum' value:'1024' type:''
update: id:0 key:'clock.force-rate' value:'44100' type:''

It looks like musescore makes a lot of effort by its own in resampling audio data probably several times from 44100 to 48000 and it seems to be more efficient just to forward the 44100 - data to pipewire and let the resampling work do the system outside of musescore. 

It looks like there had been a working version previously and we got a regression.

Edit: Just to change this one setting works fine for a few minutes. Musescore seems to reset something afterwards (or gets into memory problems), but restarting restores the ability to playback even more voices simultaneously.

**Comment by MarcSabatella** [comment]
A number of things have changed in the years since this issue was initially submitted,  it perhaps most importantly, you can now set the sample rate directly within MuseScore.  So by setting it to 48000, that should fix things in your system without altering your pipewire settings.

If for whatever reason you do wish to play with your pipewire settings instead, it is better to add 441000 to the allowed rates than to force the rate to 441000.  That lets pipewire continue to work at 48000 for other apps while also letting MuseScore set it to 441000 for its own use.

**Comment by sebicodes** [comment]
Interestingly adding 44100 to the allowed rates in pipewire doesn't solve the problem - but yes, setting the sample rate in Musescore to 48000 seems to fix the stuttering. For it took me quite a long time to find out about this, and the behaviour changed from version to version it seems still interesting to fix that problem more easily. (I don't know the reason why there were some versions in 4.6 that didn't show the stuttering).

Edit: Even changing the settings in musescore fixes the issue for just some minutes (about 9'20"). Can be fixed again by switching to 44100 and back to 48000 again without restarting musescore.
In a second try it is also about 9 minutes after startup that it doesn't work any more - turning debug mode on gives no information concerning audio-system.

Maybe it is important that switching the sampling rate initializes FluidSynth, that could be the place where the problems lie.

**Comment by jamshark70** [comment]
Update: When I use MuseScore-Studio-4.5.0.250460632-x86_64TB.AppImage (containing the fix), there's no problem.

When I use MuseScore-Studio-4.6.5.253511702-x86_64.AppImage, the issue recurs.

Was this fix never merged into the main branch?

**Comment by MarcSabatella** [comment]
Earlier in https://github.com/musescore/MuseScore/pull/23939#issuecomment-3609911699 you said it worked fine in 4.6.4.  Are you saying it broke for you between then and 4.6.5?  Does it reoccur if you try 4.6.4 again?  That seems unlikely.  It seems more likely that something has changed in your audio configuration recently - an update to pipewire, or switching to ALSA, or changing your pipewire sample rate, etc.  What settings do you see in Edit / Preferences / Audio & MIDI?  And what's your current system audio configuration?

Also, have you tried a nightly build of 4.7?

**Comment by jamshark70** [comment]
> Earlier in https://github.com/musescore/MuseScore/pull/23939#issuecomment-3609911699 you said it worked fine in 4.6.4. Are you saying it broke for you between then and 4.6.5?

Or, it's a sporadic problem, so I may simply not have encountered it that day.

> It seems more likely that something has changed in your audio configuration recently - an update to pipewire, or switching to ALSA

Hm, that could be. "Preferences" does show ALSA, which I wouldn't have set by myself. I'll try again now with Pipewire in 4.6.5 and report back later.
